File: c++/c23_2_linear_recurrences.cpp
#include <bits/stdc++.h>
#include <gtest/gtest.h>
using namespace std;
using ll = long long;
#include "c23_1_operations.h"

namespace {
	ll fib(ll n) {
		if (n <= 0)
			return 0;
		return ll(M({{0,1},{1,1}}).pow(n - 1).v[1][1]);
	}
}

TEST(C232LinearRecurrences, fib) {
	EXPECT_EQ(fib(0), 0);
	EXPECT_EQ(fib(7), 13);
}

File: python/c23_2_linear_recurrences.py
import unittest
from c23_1_operations import M

def fib(n):
    return (M([[0,1],[1,1]])**(n-1)).v[1][1] if n>0 else 0

class Test(unittest.TestCase):
    def test(self):
        self.assertEqual(fib(0),0)
        self.assertEqual(fib(7),13)

if __name__=='__main__':
    unittest.main()
